tests/smoke/alp_zgeqrf.cpp

```diff
@@ -0,0 +1,301 @@
+			if( rc != SUCCESS ) {
+				std::cerr << " dot( alpha, vi, vj, ring ) failed\n";
+				return PANIC;
+			}
+			if( i == j ) {
+				if( std::abs( *alpha - ring.template getOne< T >() ) > tol ) {
+					std::cerr << " vector " << i << " not normalized\n";
+					return PANIC;
+				}
+			} else {
+				if( std::abs( *alpha ) > tol ) {
+					std::cerr << " vector " << i << " and vctor " << j << " are note orthogonal\n";
+					return PANIC;
+				}
+			}
+#ifdef DEBUG
+			std::cout << "\t" << std::abs( *alpha );
+#endif
+		}
+#ifdef DEBUG
+		std::cout << "\n";
+#endif
+	}
+#ifdef DEBUG
+	std::cout << "\n";
+#endif
+	return rc;
+}
+
+
+//** check solution by calculating H-QR */
+template<
+	typename D,
+	typename StructureGen,
+	typename StructureOrth,
+	class Minus = operators::subtract< D >,
+	class Ring = Semiring< operators::add< D >, operators::mul< D >, identities::zero, identities::one >
+>
+RC check_solution(
+	alp::Matrix< D, StructureGen, alp::Density::Dense > &H,
+	alp::Matrix< D, StructureOrth, alp::Density::Dense > &Q,
+	alp::Matrix< D, StructureGen, alp::Density::Dense > &R,
+	const Ring &ring = Ring(),
+	const Minus &minus = Minus()
+) {
+	RC rc = SUCCESS;
+	const size_t n = nrows( H );
+	const size_t m = ncols( H );
+
+#ifdef DEBUG
+	std::cout << " ** check_solution **\n";
+	std::cout << " input matrices:\n";
+	print_matrix( " << H >> ", H );
+	print_matrix( " << Q >> ", Q );
+	print_matrix( " << R >> ", R );
+	std::cout << " ********************\n";
+#endif
+
+ 	alp::Matrix< D, StructureGen, alp::Density::Dense > QR( n, m );
+	// QR = Q * R
+	const Scalar< D > zero( ring.template getZero< D >() );
+	rc = rc ? rc : set( QR, zero );
+	rc = rc ? rc : mxm( QR, Q, R, ring );
+	// QR = QR - H
+	rc = foldl( QR, H, minus );
+
+#ifdef DEBUG
+	print_matrix( " << QR - H >> ", QR );
+#endif
+
+	//Frobenius norm
+	D fnorm = ring.template getZero< D >();
+	rc = rc ? rc : alp::eWiseLambda(
+		[ &fnorm, &ring ]( const size_t i, const size_t j, D &val ) {
+			(void) i;
+			(void) j;
+			internal::foldl( fnorm, val * val, ring.getAdditiveOperator() );
+		},
+		QR
+	);
+	fnorm = std::sqrt( fnorm );
+
+#ifdef DEBUG
+	std::cout << " FrobeniusNorm(H-QR) = " << std::abs( fnorm ) << "\n";
+#endif
+	if( tol < std::abs( fnorm ) ) {
+		std::cout << "The Frobenius norm is too large.\n";
+		return FAILED;
+	}
+
+	return rc;
+}
+
+
+
+void alp_program( const size_t & unit, alp::RC & rc ) {
+	rc = SUCCESS;
+
+	alp::Semiring<
+		alp::operators::add< ScalarType >,
+		alp::operators::mul< ScalarType >,
+		alp::identities::zero,
+		alp::identities::one
+	> ring;
+
+	// dimensions of sqare matrices H, Q and R
+	size_t N = unit;
+	size_t M = 2 * unit;
+
+	alp::Matrix< ScalarType, Orthogonal > Q( N );
+	alp::Matrix< ScalarType, General > R( N, M );
+	alp::Matrix< ScalarType, General > H( N, M );
+	{
+		auto matrix_data = generate_rectangular_matrix_data< ScalarType >( N, M );
+		rc = rc ? rc : alp::buildMatrix( H, matrix_data.begin(), matrix_data.end() );
+	}
```

Comment:
Why is this in a separate scope?

Reply:
to call deconstructor for `std::vector matrix_data`    :)